WordGraph/WordGraph.1
.\"Modified from man(1) of FreeBSD, the NetBSD mdoc.template, and mdoc.samples.
.\"See Also:
.\"man mdoc.samples for a complete listing of options
.\"man mdoc for the short list of editing options
.\"/usr/share/misc/mdoc.template
.Dd 10/16/14               \" DATE 
.Dt WordGraph 1      \" Program name and manual section number 
.Os Darwin
.Sh NAME                 \" Section Header - required - don't modify 
.Nm WordGraph,
.\" The following lines are read in generating the apropos(man -k) database. Use only key
.\" words here as the database is built based on the words here and in the .ND line. 
.Nm Other_name_for_same_program(),
.Nm Yet another name for the same program.
.\" Use .Nm macro to designate other names for the documented program.
.Nd This line parsed for whatis database.
.Sh SYNOPSIS             \" Section Header - required - don't modify
.Nm
.Op Fl abcd              \" [-abcd]
.Op Fl a Ar path         \" [-a path] 
.Op Ar file              \" [file]
.Op Ar                   \" [file ...]
.Ar arg0                 \" Underlined argument - use .Ar anywhere to underline
arg2 ...                 \" Arguments
.Sh DESCRIPTION          \" Section Header - required - don't modify
Use the .Nm macro to refer to your program throughout the man page like such:
.Nm
Underlining is accomplished with the .Ar macro like this:
.Ar underlined text .
.Pp                      \" Inserts a space
A list of items with descriptions:
.Bl -tag -width -indent  \" Begins a tagged list 
.It item a               \" Each item preceded by .It macro
Description of item a
.It item b
Description of item b
.El                      \" Ends the list
.Pp
A list of flags and their descriptions:
.Bl -tag -width -indent  \" Differs from above in tag removed 
.It Fl a                 \"-a flag as a list item
Description of -a flag
.It Fl b
Description of -b flag
.El                      \" Ends the list
.Pp
.\" .Sh ENVIRONMENT      \" May not be needed
.\" .Bl -tag -width "ENV_VAR_1" -indent \" ENV_VAR_1 is width of the string ENV_VAR_1
.\" .It Ev ENV_VAR_1
.\" Description of ENV_VAR_1
.\" .It Ev ENV_VAR_2
.\" Description of ENV_VAR_2
.\" .El                      
.Sh FILES                \" File used or created by the topic of the man page
.Bl -tag -width "/Users/joeuser/Library/really_long_file_name" -compact
.It Pa /usr/share/file_name
FILE_1 description
.It Pa /Users/joeuser/Library/really_long_file_name
FILE_2 description
.El                      \" Ends the list
.\" .Sh DIAGNOSTICS       \" May not be needed
.\" .Bl -diag
.\" .It Diagnostic Tag
.\" Diagnostic informtion here.
.\" .It Diagnostic Tag
.\" Diagnostic informtion here.
.\" .El
.Sh SEE ALSO 
.\" List links in ascending order by section, alphabetically within a section.
.\" Please do not reference files that do not exist without filing a bug report
.Xr a 1 , 
.Xr b 1 ,
.Xr c 1 ,
.Xr a 2 ,
.Xr b 2 ,
.Xr a 3 ,
.Xr b 3 
.\" .Sh BUGS              \" Document known, unremedied bugs 
.\" .Sh HISTORY           \" Document history if command behaves in a unique manner

WordGraph/wordgraph.cpp
//
//  wordgraph.cpp
//  WordGraph
//
//  Created by Anubhav on 10/16/14.
//  Copyright (c) 2014 AnubhavAshok. All rights reserved.
//

#include "wordgraph.h"
#include <boost/format.hpp>
#include <boost/optional.hpp>
#include <boost/algorithm/string.hpp>
#include <boost/filesystem.hpp>
#include <json/json.h>
#include <fstream>

using namespace std;

void WordGraph::printGraph()
{
    for (auto graphPair: wordGraph) {
        int idx = graphPair.first;
        // std::pair<std::map<int, int>, std::map<int, int>>
        auto val = graphPair.second;
        auto prev = val.first;
        auto next = val.second;
        for (auto p: prev) {
            std::string word = wordVector[p.first];
            int count = p.second;
            cout << (boost::format("(%s, %d)") % word % count).str();
        }
        if (prev.size() > 0) {
            cout << "--->";
        }
        cout << wordVector[idx];
        if (next.size() > 0) {
            cout << "<---";
        }
        for (auto p: next) {
            std::string word = wordVector[p.first];
            int count = p.second;
            cout << (boost::format("(%s, %d)") % word % count).str();
        }
        cout << endl << "-******-" << endl;
    }
}

void WordGraph::indexDoc(std::string docString)
{
    std::vector<std::string> strs;
    // Split strings by space or new line
    boost::split(strs, docString, boost::is_any_of(" \n"));
    boost::optional<int> prev;
    
    for (std::string word: strs) {
        
        int idx = 0;
        
        // Add word to wordVector and wordIndex
        if (!wordIndex.count(word)) {
            idx = (int) wordVector.size();
            wordVector.push_back(word);
            wordIndex[word] = idx;
        } else {
            idx = wordIndex[word];
        }
        
        // Set links
        if (prev) {
            int prevIdx = *prev;
            
            // set current word's prev to prev
            std::map<int, int>& myPrev = wordGraph[idx].first;
            if(!myPrev.count(prevIdx)) {
                myPrev[prevIdx] = 1;
            } else {
                myPrev[prevIdx]++;
            }
            
            // set prev's next to current word
            std::map<int, int>& prevsNext = wordGraph[prevIdx].second;
            if(!prevsNext.count(idx)) {
                prevsNext[idx] = 1;
            } else {
                prevsNext[idx]++;
            }
        }
        prev = idx;
    }
}

void WordGraph::outputToOrientDBFormat(boost::filesystem::path p)
{
    /*
    {
        nodes:[
            {
                "word": "Mary",
                "outgoing": [
                    {"type": "prev", "word": "a", "count":1},
                    {"type": "next", "word": "had", "count"},
                ]
            }
        ]
     }
     
     */
    
    
    ofstream out(p.string());
    Json::Value value;
    Json::Value nodes;
    for (std::string word: wordVector) {
        int idx = wordIndex[word];
        auto wordNode = wordGraph[idx];
        
        Json::Value node;
        node["word"] = word;
        
        Json::Value outgoing;
        // Add all prev relationships
        for (auto p: wordNode.first) {
            int idx = p.first;
            std::string prevWord = wordVector[idx];
            int count = p.second;
            Json::Value rel;
            rel["type"] = "prev";
            rel["word"] = prevWord;
            rel["count"] = count;
            outgoing.append(rel);
        }
        
        // Add all next relationships
        for (auto p: wordNode.second) {
            int idx = p.first;
            std::string nextWord = wordVector[idx];
            int count = p.second;
            Json::Value rel;
            rel["type"] = "next";
            rel["word"] = nextWord;
            rel["count"] = Json::Value(count);
            outgoing.append(rel);
        }
        
        node["outgoing"] = outgoing;
        nodes.append(node);
    }
    value["nodes"] = nodes;

    
    out << value << endl;
}
